File: Data Structures & Algorithms/subsets/submission-2.java
class Solution {
    Set<List<Integer>> res;

    private void compute(int[] nums, int index, Deque<Integer> curr) {
        if (index == nums.length) {
            res.add(new ArrayList<>(curr));
            return;
        }
        curr.push(nums[index]);
        compute(nums, index+1, curr);
        curr.pop();
        compute(nums, index+1, curr);
    }
    public List<List<Integer>> subsets(int[] nums) {
       // sort of a decision tree
       res = new HashSet<>();
       Deque<Integer> queue = new ArrayDeque<>();
       compute(nums, 0, queue);
       return new ArrayList<>(res);
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-2.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = set()
        nums.sort()
        def subsets(i, curr):
            if i == len(nums):
                res.add(tuple(curr.copy()))
                return
            curr.append(nums[i])
            subsets(i+1, curr)
            curr.pop()
            subsets(i+1, curr)
        subsets(0, [])
        groups = []
        for group in res:
            groups.append(list(group))
        return groups
        